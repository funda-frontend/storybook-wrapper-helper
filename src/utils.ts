import { camelCase } from 'change-case';
import { LayoutOptions, SelectOption } from './types';

/**
 * Pages should not have a padding around the edge to more closely represent
 * what the page is going to look like. This can be overwritten at the story
 * level.
 */
export const defaultPageConfigurationValues = {
    layout: LayoutOptions.FullScreen,
};

export function isInPageFolder(title: string) {
    return title.toLowerCase().slice(0, 4) === 'page';
}

export function generateFunctionName(input) {
    return camelCase(input);
}

/**
 * Storybook always needs a template for vue components, but in most cases, they
 * are just going to be the name of the component and include bindings for props
 * and events. This function generates these basic templates.
 */
export function generateTemplate({
    componentName,
    events = [],
}: {
    componentName: string;
    events?: Array<string>;
}): string {
    const eventsString = events
        .map((event) => `@${event}="${generateFunctionName(event)}"`)
        .join(' ');

    let str = `<${componentName} v-bind="$props"`;
    if (eventsString) {
        str += ` ${eventsString}`;
    }
    str += '/>';

    return str;
}

/**
 * Generate a markdown section for links that will open in a new tab when clicked
 */
export function markdownLinks(links: Array<SelectOption>): string {
    const linksHeader = '\n### Links\n';
    const linksBody = links
        .map(
            (link) =>
                `* <a href="${link.value}" target="_blank">${link.name}</a>`
        )
        .join('\n');
    return linksHeader + linksBody;
}

/** Transform a string generator into the template storybook requires */
export function generateDecorator(decoratorString) {
    return () => ({
        template: decoratorString,
    });
}

export function objectHasContent(obj) {
    return obj && Object.keys(obj).length;
}

export function arrayHasContent(arr) {
    return arr && arr.length;
}
