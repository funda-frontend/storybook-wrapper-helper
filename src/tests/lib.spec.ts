import { generateTemplate } from '../utils';
import { book, storyFunctionPropsToStoryProps } from '../lib';

const ExampleObject = { example: 'object' };
const AdditionalComponent = { additional: 'component' };

describe('book', () => {
    const defaultBookProps = { component: { ExampleObject } };

    const setup = (args = {}) =>
        book({
            ...defaultBookProps,
            ...args,
        });

    it('should transform component input', () => {
        const result = setup();
        expect(result.component).toEqual(ExampleObject);
        expect(result.componentName).toEqual('ExampleObject');
    });

    it('should generate proper decorators', () => {
        const decoratorString = '<div><story/></div>';

        const bookSettings = setup({
            decorators: [decoratorString],
        });

        expect((bookSettings.decorators as any)[0]()).toEqual({
            template: decoratorString,
        });
    });

    it('should bind events', () => {
        const events = ['close'];
        const result = setup({ events });

        expect(result.argTypes).toEqual({ close: { action: 'close' } });
        expect(result.events).toEqual(events);
    });

    it('should add a description', () => {
        const description = 'hello world';
        const result = setup({ description });

        expect(result.parameters?.docs?.description?.component).toEqual(
            description
        );
    });

    it('should add links when a description is not present', () => {
        const url = 'https://example.com';
        const result = setup({
            links: [{ name: 'link', value: url }],
        });

        expect(result.parameters?.docs?.description?.component).toContain(
            `<a href=\"${url}\"`
        );
    });
});

describe('storyFunctionPropsToStoryProps', () => {
    const baseStoryFunctionProps = {
        componentName: 'ExampleObject',
        component: ExampleObject,
    };

    const setup = (args = {}) =>
        storyFunctionPropsToStoryProps({ ...baseStoryFunctionProps, ...args });

    it('should define a template and components', () => {
        const result = setup();

        expect(result.components).toEqual({ ExampleObject });
        expect(result.template).toEqual(
            generateTemplate({
                componentName: 'ExampleObject',
            })
        );
    });

    it('should overwrite the default template with a custom one', () => {
        const customTemplate = '<ExampleObject/>';
        const result = setup({
            template: customTemplate,
        });

        expect(result.template).toEqual(customTemplate);
    });

    it('should allow for additional components', () => {
        const result = setup({
            additionalComponents: {
                AdditionalComponent,
            },
        });

        expect(Object.keys(result.components)).toContain('AdditionalComponent');
    });

    it('should add a description', () => {
        const description = 'hello world';
        const result = setup({ description });

        expect(result.parameters?.docs?.description?.story).toEqual(
            description
        );
    });
});

describe('snapshot tests', () => {
    const bookSettings = book({
        title: 'Component/Path',
        args: { key: 'value' },
        component: { ExampleObject },
        decorators: ['<div><story/></div>'],
        description: 'hello world',
        links: [{ name: 'link', value: 'https://example.com' }],
        events: ['close'],
        template: `
            <ExampleObject v-bind="$props" @close="close">
                Slot Content
            </ExampleObject>`,
        argTypes: {
            showPrefix: {
                control: {
                    type: 'boolean',
                },
            },
            objectProp: {
                description: 'Documentation for an opaque object structure',
            },
        },
    });
    it('should match the expected book results', () => {
        expect(bookSettings).toMatchSnapshot();
    });
    it('should match the story props results', () => {
        const result = storyFunctionPropsToStoryProps({
            ...bookSettings,
            additionalComponents: { AdditionalComponent },
            decorators: ['<p><story/></p>'],
            description: 'hello world',
            links: [{ name: 'link', value: 'https://example.com' }],
            parameters: {
                backgrounds: {
                    default: 'custom-blue',
                    values: [
                        {
                            name: 'custom-blue',
                            value: '#139ec1',
                        },
                    ],
                },
            },
        });
        expect(result).toMatchSnapshot({});
    });
});
