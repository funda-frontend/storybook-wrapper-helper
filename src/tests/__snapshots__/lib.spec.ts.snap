// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot tests should match the expected book results 1`] = `
Object {
  "argTypes": Object {
    "close": Object {
      "action": "close",
    },
    "objectProp": Object {
      "description": "Documentation for an opaque object structure",
    },
    "showPrefix": Object {
      "control": Object {
        "type": "boolean",
      },
    },
  },
  "component": Object {
    "example": "object",
  },
  "componentName": "ExampleObject",
  "decorators": Array [
    [Function],
    [Function],
  ],
  "defaultArgs": Object {
    "key": "value",
  },
  "events": Array [
    "close",
  ],
  "parameters": Object {
    "design": Array [
      Object {
        "type": "figma",
        "url": "https://www.figma.com/file/file-key",
      },
      Object {
        "type": "link",
        "url": "https://www.figma.com/file/file-key-2",
      },
    ],
    "docs": Object {
      "description": Object {
        "component": "hello world
### Links
* <a href=\\"https://example.com\\" target=\\"_blank\\">link</a>",
      },
    },
  },
  "template": "
            <ExampleObject v-bind=\\"$props\\" @close=\\"close\\">
                Slot Content
            </ExampleObject>",
  "title": "Component/Path",
}
`;

exports[`snapshot tests should match the story props results 1`] = `
Object {
  "argTypes": Object {
    "close": Object {
      "action": "close",
    },
    "objectProp": Object {
      "description": "Documentation for an opaque object structure",
    },
    "showPrefix": Object {
      "control": Object {
        "type": "boolean",
      },
    },
  },
  "args": Object {
    "item": 1,
    "key": "value",
  },
  "components": Object {
    "AdditionalComponent": Object {
      "additional": "component",
    },
    "ExampleObject": Object {
      "example": "object",
    },
  },
  "decorators": Array [
    [Function],
    [Function],
  ],
  "links": Array [
    Object {
      "name": "link",
      "value": "https://example.com",
    },
  ],
  "parameters": Object {
    "backgrounds": Object {
      "default": "custom-blue",
      "values": Array [
        Object {
          "name": "custom-blue",
          "value": "#139ec1",
        },
      ],
    },
    "design": Object {
      "type": "figma",
      "url": "https://www.figma.com/file/file-key-3",
    },
    "docs": Object {
      "description": Object {
        "story": "hello world
### Links
* <a href=\\"https://example.com\\" target=\\"_blank\\">link</a>",
      },
    },
  },
  "template": "
            <ExampleObject v-bind=\\"$props\\" @close=\\"close\\">
                Slot Content
            </ExampleObject>",
  "title": "Component/Path",
}
`;
